designs/setdict-design.md

```diff
@@ -0,0 +1,152 @@
+# Set/Tuple/Dictionary: Project Design
+
+## Week 7 Update
+
+### Passed Test Cases:
+1. Initialize the set, and measure the length of the set using len().
+```python
+set_1 : set[int] = None
+set_1 = {1,2}
+print(len(set_1)) # output 2
+```
+
+2. Add elements to the set.
+```python
+set_1 : set[int] = None
+set_1 = {1,2}
+set_1.add(3)
+print(len(set_1)) # output 3
+```
+
+3. Use the "in" keyword to see if an element exists in the set.
+```python
+set_1 : set[int] = None
+set_1 = {1,2}
+set_1.add(3)
+print(3 in set_1) # output True
+```
+
+4. "in" should return False if element doesn't exist.
+```python
+set_1 : set[int] = None
+set_1 = {1,2}
+set_1.add(3)
+print(4 in set_1) # output False
+```
+
+5. Duplicate elements will not be inserted into the set again.
+```python
+set_1 : set[int] = None
+set_1 = {1,2}
+set_1.add(1)
+print(len(set_1))  # output 2   
+```
+
+6. Use remove() to remove an element from the set.
+```python
+set_1 : set[int] = None
+set_1 = {1,2}
+set_1.remove(1)
+print(len(set_1)) # output 1 
+```
+
+7. "in" should return False after removing
+```python
+set_1 : set[int] = None
+set_1 = {1,2}
+set_1.remove(1)
+print(1 in set_1) # output False
+```
+
+
+Some functionalities remaining: set() constructor; remove() non existing element should throw runtime error; clear(); update() (should include iterable -- string, list, set...), copy().
+
+Initially we wanted to implement the set using a linear memory layout, but that still requires writing wasm code for searching for duplicates while adding to the set, yielding O(n) time complexity. We then switched to a hash-based solution, implementing a hash table in wasm. We currently use mod10 as our hash function, and therefore give out 10 buckets in the table. Each bucket is initially assigned 4 bytes to store the starting address and maintains a linked list. The functions such as add(), remove(), "in" all perform similarly: use the hash function to find the correct bucket, then iterate through the linked list to perform further actions.
+
+In the following weeks, we expect to 1. provide support for the remaining functionalities and behaviors of set; 2. extend our implementation to work with strings and lists; 3. if there's enough time, we will also try to implement tuple and dictionary.

```

Comment:
What does working with strings and lists mean?

Reply:
Actually, set should be worked with string and lists in some functions such as set.update and set initialization, too. Now, our work does not involve the string and lists, we would finish these work after communicating with string and list group.

Reply:
I understand that you probably expect a more detailed explanation than "working with", and we will include it in our updated design.md.